### DISCUSSION TITLE
Driving backwards vs Custom dock modification

### DISCUSSION THREAD

**Discussion by martincerven** [iRobotEducation/create3_docs#620] [Questions]
I'd like to attach mechanical device in front of robot to perform object manipulation. I used this approach with Kobuki (Turtlebot 2) with [no problems](https://youtu.be/ZjAa_kfo6S0). With Create 3 there is dock (Home base) that prevents this since it is only way of charging the robot. 

Is there a way to just attach anything to the back of the robot and flip forward/backward directions? I'm using ROS navigation so I wouldn't be using bumper & bottom IR sensors anyways. I can think of few issues:

- different wheel dynamics in forward/backward directions?
- stability: could be solved by attaching additional caster at the back (now front)
- firmware on intergrated board don't support this and would need to be modified

If the forward/backward flip could be solved within ROS without touching embedded board that'd be cool.

**Another** way this could be solved would be 3D printing custom dock, a flat one, where robot would just drive onto it. Issues could be:

- proprietary iRobot HW inside dock (charging system)
- would need huge 3D printer 
- would need special power supply

I think modifying Create3 in this way would be huge benefit for community and owners of Create3 robot. Does anyone have experience with either flipping directions of Create3 ( or previous Turtlebots ) and/or creating custom dock/charging?

**Comment by alsora** [comment]
> 3D printing custom dock, a flat one, where robot would just drive onto it

The dock does two main things: 1) you have the contacts that charge the robot and 2) you have IR emitters that allow the robot to see the dock (that's how the `Dock` ROS 2 action is implemented on the create 3 robot.
3D printing a custom dock that acts as a drop-in replacement for the iRobot one may be quite complex


> Is there a way to just attach anything to the back of the robot and flip forward/backward directions?

I'm not sure I understand this part.
If you attach something to the back, then you should be good.
The robot will be able to dock normally.

If you mean that you want to attach something to the front, and you are asking if the robot can dock backward that's a different story.
 - The on-board algorithm can't be modified to dock backward
 - The standard dock doesn't allow docking backward because the battery contacts are on the front of the robot.

I'm not a battery expert, but assuming that you can charge the Create 3 battery without using the official dock, you could design a new one and run your own docking algorithm (for example the nav2 stack has some open-source apriltag based docking solutions).

**Comment by martincerven** [comment]
There was addition of [Docking Server](https://docs.nav2.org/configuration/packages/configuring-docking-server.html#docking-server) to nav2 by Steve Macenski recently, so navigation to custom dock could be supported by ROS2.

> I'm not a battery expert, but assuming that you can charge the Create 3 battery without using the official dock, you could design a new one and run your own docking algorithm (for example the nav2 stack has some open-source apriltag based docking solutions).

This is one part of the question, there is for sure circuit in Home Base dock which detects:

-  if there is no robot present: voltage on charging pads is ~ 3.3V 
-  when there is robot, it is increased to ~17V -  make sense for safety.

On Kobuki you just plugged power supply with barrel jack 18V DC in addition to dock (which I didn't have).
So maybe you could provide 17V DC to charging pads at the bottom and robot's battery would charge? ( _Is there BMS on robot or is this dock's functionality?_ )

I  have battery hacking in my todo-list for future, but for battery packs there are **BMS** (Battery management system) that protect cells from overcharging and manages each cell's voltage.

My second part of question is, if I wanted to keep using dock for charging, to attach something in the back of the robot (with caster), and use all ROS functionality (cmd_vel, nav2) with forward/backward flipped? 

Both parts are kinda hacky, in short, I just want to attach some mechanism _close to the floor in front of the robot,_ but there is dock in the way, so either design custom (flat) dock/charger or put the mechanism at the back, and flip directions software wise (while ignoring physical implications).

3rd option is to use some retractable mechanism that would lower itself down while not charging but that is even more sci-fi ( _Would look cool though_)

What do you think is most approachable way?

**Reply by martincerven** [reply]
For illustration:![image](https://github.com/user-attachments/assets/f54a1488-c8da-4990-b562-bd9c3c841c4d)

**Reply by alsora** [reply]
Ok got it.
P.S. Your robots are very cool.

> So maybe you could provide 17V DC to charging pads at the bottom and robot's battery would charge? ( Is there BMS on robot or is this dock's functionality? )

I'll need to check with someone familiar with the battery.

> attach something in the back of the robot (with caster), and use all ROS functionality (cmd_vel, nav2) with forward/backward flipped

This might work.
It shouldn't require any change to the create 3 Firmware code.
You would just provide the velocities flipped from your app or the nav stack.
Make sure to disable safety backup from the ROS 2 safety parameters.

**Reply by Mechazo11** [reply]
Hi @alsora I know this is old post but was wondering which file contains the ROS 2 safety parameter to disable backup drive?

**Comment by martincerven** [comment]
@alsora Did you get any info about battery/direct charging? Thanks!
